qiskit/transpiler/preset_passmanagers/common.py

```diff
@@ -51,6 +52,96 @@
 from qiskit.transpiler.exceptions import TranspilerError
 from qiskit.transpiler.layout import Layout
 
+_CONTROL_FLOW_OP_NAMES = {"for_loop", "if_else", "while_loop"}
+
+# Structing is {type: (known good, known bad)}.  Anything neither known good nor known bad (i.e.
+# not a Terra-internal pass) is permitted to pass through without error, since it is being supplied
+# by a plugin and we don't have any knowledge of these.
+_CONTROL_FLOW_STATES = {
+    "layout_method": ({"trivial", "dense"}, {"sabre", "noise_adaptive"}),
+    "routing_method": ({"none", "stochastic"}, {"sabre", "lookahead", "basic", "toqm"}),
+    # 'synthesis' is not a supported translation method because of the block-collection passes
+    # involved; we currently don't have a neat way to pass the information about nested blocks - the
+    # `UnitarySynthesis` pass itself is control-flow aware.
+    "translation_method": ({"translator", "unroller"}, {"synthesis"}),
+    "optimization_method": (set(), set()),
+    "scheduling_method": (set(), {"alap", "asap"}),
+}
+
+
+def _has_control_flow(property_set):
+    return any(property_set[f"contains_{x}"] for x in _CONTROL_FLOW_OP_NAMES)
+
+
+def _without_control_flow(property_set):
+    return not any(property_set[f"contains_{x}"] for x in _CONTROL_FLOW_OP_NAMES)
+
+
+def generate_control_flow_options_check(
+    layout_method=None,
+    routing_method=None,
+    translation_method=None,
+    optimization_method=None,
+    scheduling_method=None,
+):
+    """Generate a pass manager that, when run on a DAG that contains control flow, fails with an
+    error message explaining the invalid options, and what could be used instead.
+
+    Returns:
+        PassManager: a pass manager that populates the ``contains_x`` properties for each of the
+        control-flow operations, and raises an error if any of the given options do not support
+        control flow, but a circuit with control flow is given.
+    """
+
+    bad_options = []
+    message = "Some options cannot be used with control flow."
+    for stage, given in [
+        ("layout", layout_method),
+        ("routing", routing_method),
+        ("translation", translation_method),
+        ("optimization", optimization_method),
+        ("scheduling", scheduling_method),
+    ]:
+        option = stage + "_method"
```

Comment:
I think I did this just because one of the possible error messages is "the entire {stage} stage is not supported", and that's weird with `scheduling_method`.